src/AutoRest.CSharp/Mgmt/Output/Models/PolymorphicOption.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using AutoRest.CSharp.Mgmt.Decorator;
+using AutoRest.CSharp.Mgmt.Models;
+using AutoRest.CSharp.Output.Models;
+using AutoRest.CSharp.Output.Models.Shared;
+using Azure.Core;
+
+namespace AutoRest.CSharp.Mgmt.Output.Models;
+
+internal class PolymorphicOption
+{
+    public Resource Resource { get; }
+
+    public BaseResource BaseResource { get; }
+
+    public ScopeResourceTypeConstraint? ScopeResourceTypeConstraint { get; }
+
+    public Segment? ResourceNameConstraint { get; }
+
+    public PolymorphicOption(Resource resource, BaseResource baseResource)
+    {
+        Resource = resource;
+        BaseResource = baseResource;
+        var requestPath = resource.RequestPath;
+        var scopePath = requestPath.GetScopePath();
+        var scopeResourceType = scopePath.GetResourceType();
+        if (scopeResourceType != ResourceTypeSegment.Any && scopeResourceType != ResourceTypeSegment.Scope)
+        {
+            // get how many segments we have for those "non-scope" parts
+            var trimmed = requestPath.TrimScope();
+            Debug.Assert(trimmed.Count % 2 == 0); // this is ensured by our resource detection logic -- might throws if there is a manual resource data configuration which assigns a path with odd segments as resource
+            var countOfParents = trimmed.Count / 2;
+            if (trimmed.First().Equals(Segment.Providers))
+                countOfParents--;
+            var builder = new StringBuilder();
+            for (int i = 0; i < countOfParents; i++)
+            {
+                builder.Append(".Parent");
+            }
+            ScopeResourceTypeConstraint = new ScopeResourceTypeConstraint($"{IdParameter.Name}{builder.ToString()}", scopeResourceType);
+        }
+        var resourceNameSegment = requestPath.Last();
+        if (resourceNameSegment.IsConstant)
+            ResourceNameConstraint = resourceNameSegment;
+    }
+
+    private Parameter? _idParameter;
+    public Parameter IdParameter => _idParameter ??= new Parameter(Name: "id", Description: null, Type: typeof(ResourceIdentifier), DefaultValue: null, Validation: ValidationType.AssertNotNull, Initializer: null);
```

Comment:
Why this one can't be instantiated in constructor?

Reply:
Actually, why this one can't be static?

Reply:
I made it readonly static